Implement tensor.select(Dimname,int)

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21919 Added names= arg for some creation functions
* #22077 Refactor function_wrapper.create_generic
* **#21795 Implement tensor.select(Dimname,int)**

tensor.select(Dimname, int) wraps tensor.select(int,int), where the
named inference rules happen.

Codegen changes:
- named tensor code in core/{Type.h, Tensor.h, TensorMethods.h} is
guarded with macros.

Test Plan
- `python test/test_namedtensor.py -v` [namedtensor ci]

gh-metadata: pytorch pytorch 21795 gh/zou3519/54/head

Differential Revision: [D15833457](https://our.internmc.facebook.com/intern/diff/D15833457)

zou3519 committed:
Implement tensor.select(Dimname,int)

tensor.select(Dimname, int) wraps tensor.select(int,int), where the
named inference rules happen.

Test Plan
- `python test/test_namedtensor.py -v` [namedtensor ci]

zou3519 committed:
Update on "Implement tensor.select(Dimname,int)"

Implement tensor.select(Dimname,int)

tensor.select(Dimname, int) wraps tensor.select(int,int), where the
named inference rules happen.

Test Plan
- `python test/test_namedtensor.py -v` [namedtensor ci]

gh-metadata: pytorch pytorch 21795 gh/zou3519/54/head

zou3519 committed:
Update on "Implement tensor.select(Dimname,int)"

Implement tensor.select(Dimname,int)

tensor.select(Dimname, int) wraps tensor.select(int,int), where the
named inference rules happen.

Test Plan
- `python test/test_namedtensor.py -v` [namedtensor ci]

gh-metadata: pytorch pytorch 21795 gh/zou3519/54/head

zou3519 committed:
Update on "Implement tensor.select(Dimname,int)"

Implement tensor.select(Dimname,int)

tensor.select(Dimname, int) wraps tensor.select(int,int), where the
named inference rules happen.

Test Plan
- `python test/test_namedtensor.py -v` [namedtensor ci]

gh-metadata: pytorch pytorch 21795 gh/zou3519/54/head

zou3519 committed:
Update on "Implement tensor.select(Dimname,int)"

Implement tensor.select(Dimname,int)

tensor.select(Dimname, int) wraps tensor.select(int,int), where the
named inference rules happen.

Test Plan
- `python test/test_namedtensor.py -v` [namedtensor ci]

gh-metadata: pytorch pytorch 21795 gh/zou3519/54/head

zou3519 committed:
Update on "Implement tensor.select(Dimname,int)"

Implement tensor.select(Dimname,int)

tensor.select(Dimname, int) wraps tensor.select(int,int), where the
named inference rules happen.

Codegen changes:
- named tensor code in core/{Type.h, Tensor.h, TensorMethods.h} is
guarded with macros.

Test Plan
- `python test/test_namedtensor.py -v` [namedtensor ci]

gh-metadata: pytorch pytorch 21795 gh/zou3519/54/head

zou3519 committed:
Update on "Implement tensor.select(Dimname,int)"

Implement tensor.select(Dimname,int)

tensor.select(Dimname, int) wraps tensor.select(int,int), where the
named inference rules happen.

Codegen changes:
- named tensor code in core/{Type.h, Tensor.h, TensorMethods.h} is
guarded with macros.

Test Plan
- `python test/test_namedtensor.py -v` [namedtensor ci]

gh-metadata: pytorch pytorch 21795 gh/zou3519/54/head

zou3519 committed:
Update on "Implement tensor.select(Dimname,int)"

Implement tensor.select(Dimname,int)

tensor.select(Dimname, int) wraps tensor.select(int,int), where the
named inference rules happen.

Codegen changes:
- named tensor code in core/{Type.h, Tensor.h, TensorMethods.h} is
guarded with macros.

Test Plan
- `python test/test_namedtensor.py -v` [namedtensor ci]

gh-metadata: pytorch pytorch 21795 gh/zou3519/54/head

nairbv commented on tools/jit/gen_jit_dispatch.py:
```diff
@@ -171,7 +171,7 @@ def from_ivalue(arg, value):
 
 
-blacklisted_types = {'Storage', 'Dimname', 'DimnameList?'}
+blacklisted_types = {'Storage', 'DimnameList?', 'ConstQuantizerPtr', 'Dimname', 'DimnameList?'}
```
DimnameList twice here?

zou3519 replied:
Nice catch

zou3519 committed:
Update on "Implement tensor.select(Dimname,int)"

Implement tensor.select(Dimname,int)

tensor.select(Dimname, int) wraps tensor.select(int,int), where the
named inference rules happen.

Codegen changes:
- named tensor code in core/{Type.h, Tensor.h, TensorMethods.h} is
guarded with macros.

Test Plan
- `python test/test_namedtensor.py -v` [namedtensor ci]

gh-metadata: pytorch pytorch 21795 gh/zou3519/54/head

zou3519 committed:
Update on "Implement tensor.select(Dimname,int)"

Implement tensor.select(Dimname,int)

tensor.select(Dimname, int) wraps tensor.select(int,int), where the
named inference rules happen.

Codegen changes:
- named tensor code in core/{Type.h, Tensor.h, TensorMethods.h} is
guarded with macros.

Test Plan
- `python test/test_namedtensor.py -v` [namedtensor ci]

gh-metadata: pytorch pytorch 21795 gh/zou3519/54/head

zou3519 commented:
@ailzhang XLA failure. Please let me know how you would like to proceed.

ailzhang commented:
Hi @zou3519 Thanks! I think this will be a simple schema change at the first glance. We should be able to make a followup change once this is merged. cc: @dlibenzi @asuhan

dlibenzi commented:
> Hi @zou3519 Thanks! I think this will be a simple schema change at the first glance. We should be able to make a followup change once this is merged. cc: @dlibenzi @asuhan

Looks like the code which is generating `RegistrationDeclarations.h` needs fixing?

zou3519 commented:
> Hi @zou3519 Thanks! I think this will be a simple schema change at the first glance. We should be able to make a followup change once this is merged. cc: @dlibenzi @asuhan

I'm not sure if it's a schema change; I think you just need to blacklist Dimname/DimnameList in the code generation

dlibenzi commented:
> > Hi @zou3519 Thanks! I think this will be a simple schema change at the first glance. We should be able to make a followup change once this is merged. cc: @dlibenzi @asuhan
> 
> I think you just need to blacklist Dimname/DimnameList in the code generation

Why blacklisting?
Can you post the `RegistrationDeclarations.h` generated by your PR?
Or at least the incriminated line?

zou3519 commented:
> Why blacklisting?
Can you post the RegistrationDeclarations.h generated by your PR?
Or at least the incriminated line?

https://gist.github.com/zou3519/40ef0f94399d53c8996720bec852908f
I think the problem is that the NAMEDTENSOR_ENABLED macros aren't being respected somewhere downstream. I could change it so that everything that is currently wrapped inside the NAMEDTENSOR_ENABLED sections are not written to the RegistrationDeclarations.h file if that is the only file being consumed.

dlibenzi commented:
> > Why blacklisting?
> > Can you post the RegistrationDeclarations.h generated by your PR?
> > Or at least the incriminated line?
> 
> https://gist.github.com/zou3519/40ef0f94399d53c8996720bec852908f
> I think the problem is that the NAMEDTENSOR_ENABLED macros aren't being respected somewhere downstream. I could change it so that everything that is currently wrapped inside the NAMEDTENSOR_ENABLED sections are not written to the RegistrationDeclarations.h file if that is the only file being consumed.

I think, if XLA should not generate code for it, the #ifdef NAMEDTENSOR_ENABLED shouldn't even be included in that file

zou3519 commented:
> > > Why blacklisting?
> > > Can you post the RegistrationDeclarations.h generated by your PR?
> > > Or at least the incriminated line?
> > 
> > 
> > https://gist.github.com/zou3519/40ef0f94399d53c8996720bec852908f
> > I think the problem is that the NAMEDTENSOR_ENABLED macros aren't being respected somewhere downstream. I could change it so that everything that is currently wrapped inside the NAMEDTENSOR_ENABLED sections are not written to the RegistrationDeclarations.h file if that is the only file being consumed.
> 
> I think, if XLA should not generate code for it, the #ifdef NAMEDTENSOR_ENABLED shouldn't even be included in that file

Are there other files I should be aware of that XLA uses to generate code, or is this (RegistrationDeclarations) the only one?

dlibenzi commented:
> > > > Why blacklisting?
> > > > Can you post the RegistrationDeclarations.h generated by your PR?
> > > > Or at least the incriminated line?
> > > 
> > > 
> > > https://gist.github.com/zou3519/40ef0f94399d53c8996720bec852908f
> > > I think the problem is that the NAMEDTENSOR_ENABLED macros aren't being respected somewhere downstream. I could change it so that everything that is currently wrapped inside the NAMEDTENSOR_ENABLED sections are not written to the RegistrationDeclarations.h file if that is the only file being consumed.
> > 
> > 
> > I think, if XLA should not generate code for it, the #ifdef NAMEDTENSOR_ENABLED shouldn't even be included in that file
> 
> Are there other files I should be aware of that XLA uses to generate code, or is this (RegistrationDeclarations) the only one?

We also use `Functions.h` and `NativeFunctions.h`, but you can leave those alone.
The real driver for the code generation is `RegistrationDeclarations.h`.

https://github.com/pytorch/xla/blob/master/scripts/generate_code.sh

zou3519 committed:
Update on "Implement tensor.select(Dimname,int)"

Implement tensor.select(Dimname,int)

tensor.select(Dimname, int) wraps tensor.select(int,int), where the
named inference rules happen.

Codegen changes:
- named tensor code in core/{Type.h, Tensor.h, TensorMethods.h} is
guarded with macros.

Test Plan
- `python test/test_namedtensor.py -v` [namedtensor ci]

gh-metadata: pytorch pytorch 21795 gh/zou3519/54/head

dlibenzi commented:
> > > > > Why blacklisting?
> > > > > Can you post the RegistrationDeclarations.h generated by your PR?
> > > > > Or at least the incriminated line?
> > > > 
> > > > 
> > > > https://gist.github.com/zou3519/40ef0f94399d53c8996720bec852908f
> > > > I think the problem is that the NAMEDTENSOR_ENABLED macros aren't being respected somewhere downstream. I could change it so that everything that is currently wrapped inside the NAMEDTENSOR_ENABLED sections are not written to the RegistrationDeclarations.h file if that is the only file being consumed.
> > > 
> > > 
> > > I think, if XLA should not generate code for it, the #ifdef NAMEDTENSOR_ENABLED shouldn't even be included in that file
> > 
> > 
> > Are there other files I should be aware of that XLA uses to generate code, or is this (RegistrationDeclarations) the only one?
> 
> We also use `Functions.h` and `NativeFunctions.h`, but you can leave those alone.
> The real driver for the code generation is `RegistrationDeclarations.h`.
> 
> https://github.com/pytorch/xla/blob/master/scripts/generate_code.sh

We could pass that through `cpp` on our side if you like.

zou3519 commented:
> > > > > > Why blacklisting?
> > > > > > Can you post the RegistrationDeclarations.h generated by your PR?
> > > > > > Or at least the incriminated line?
> > > > > 
> > > > > 
> > > > > https://gist.github.com/zou3519/40ef0f94399d53c8996720bec852908f
> > > > > I think the problem is that the NAMEDTENSOR_ENABLED macros aren't being respected somewhere downstream. I could change it so that everything that is currently wrapped inside the NAMEDTENSOR_ENABLED sections are not written to the RegistrationDeclarations.h file if that is the only file being consumed.
> > > > 
> > > > 
> > > > I think, if XLA should not generate code for it, the #ifdef NAMEDTENSOR_ENABLED shouldn't even be included in that file
> > > 
> > > 
> > > Are there other files I should be aware of that XLA uses to generate code, or is this (RegistrationDeclarations) the only one?
> > 
> > 
> > We also use `Functions.h` and `NativeFunctions.h`, but you can leave those alone.
> > The real driver for the code generation is `RegistrationDeclarations.h`.
> > https://github.com/pytorch/xla/blob/master/scripts/generate_code.sh
> 
> We could pass that through `cpp` on our side if you like.

What are you referring to by "that"?

I fixed up RegistrationDeclarations.h in this pull request so that it doesn't emit any `#ifdef NAMEDTENSOR_ENABLED` code and it looks like the build succeeded so all should be good.

dlibenzi commented:
I meant we could pipe that through the C++ preprocessor before our parsing
... but if you already fixed that, we are good


On Fri, Jun 21, 2019, 20:01 Richard Zou <notifications@github.com> wrote:

> Why blacklisting?
> Can you post the RegistrationDeclarations.h generated by your PR?
> Or at least the incriminated line?
>
> https://gist.github.com/zou3519/40ef0f94399d53c8996720bec852908f
> I think the problem is that the NAMEDTENSOR_ENABLED macros aren't being
> respected somewhere downstream. I could change it so that everything that
> is currently wrapped inside the NAMEDTENSOR_ENABLED sections are not
> written to the RegistrationDeclarations.h file if that is the only file
> being consumed.
>
> I think, if XLA should not generate code for it, the #ifdef
> NAMEDTENSOR_ENABLED shouldn't even be included in that file
>
> Are there other files I should be aware of that XLA uses to generate code,
> or is this (RegistrationDeclarations) the only one?
>
> We also use Functions.h and NativeFunctions.h, but you can leave those
> alone.
> The real driver for the code generation is RegistrationDeclarations.h.
> https://github.com/pytorch/xla/blob/master/scripts/generate_code.sh
>
> We could pass that through cpp on our side if you like.
>
> What are you referring to by "that"?
>
> I fixed up RegistrationDeclarations.h in this pull request so that it
> doesn't emit any #ifdef NAMEDTENSOR_ENABLED code and it looks like the
> build succeeded so all should be good.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/21795?email_source=notifications&email_token=ADR5YXFZSMCTGH7D3SDCYBLP3UJQLA5CNFSM4HYLZZPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYJFGMY#issuecomment-504517427>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADR5YXD7WS2YQ4W6OUSEJPLP3UJQLANCNFSM4HYLZZPA>
> .
>
